qutip/wigner.py

```diff
@@ -548,83 +548,297 @@ def _wig_laguerre_val(L, x, c):
 
-    Returns
+    See Also
     --------
-    Q : array
-        Values representing the Q-function calculated over the specified range
-        [xvec,yvec].
-
+    :obj:`.qfunc` :
+        a single function version, which will involve computing several
+        quantities multiple times in order to use less memory.
     """
-    X, Y = meshgrid(xvec, yvec)
-    amat = 0.5 * g * (X + Y * 1j)
-
-    if not (isoper(state) or isket(state)):
-        raise TypeError('Invalid state operand to qfunc.')
-
-    qmat = zeros(size(amat))
-
-    if isket(state):
-        qmat = _qfunc_pure(state, amat)
-    elif isoper(state):
-        d, v = la.eig(state.full())
-        # d[i]   = eigenvalue i
-        # v[:,i] = eigenvector i
-
-        qmat = zeros(np.shape(amat))
-        for k in arange(0, len(d)):
-            qmat1 = _qfunc_pure(v[:, k], amat)
-            qmat += real(d[k] * qmat1)
+    def __init__(self, xs, ys, g: float = np.sqrt(2), memory: float = 1024):
+        self._g = g
+        self._alpha_matrix = _qfunc_alpha_matrix(xs, ys, g)
+        # 16 bytes per complex, 1024**2 bytes per MB.
+        self._size_mb = self._alpha_matrix.conj.size * 16 / (1024**2)
+        self._memory_mb = memory
+        self._max_size = int(self._memory_mb // self._size_mb)
+        self._current_size = 0
+        self._cache = None
+
+    def _alphas(self, size: int):
+        r"""
+        Retrive the full tensor of (the conjugate of) coherent states over all
+        values of :math:`\alpha`, for states of dimension ``size``.
+        """
+        if self._current_size >= size:
+            return self._cache[:, :, :size]
+        if size > self._max_size:
+            requirement = self._size_mb * size
+            raise MemoryError(
+                f"Refusing to precompute up to {size} basis states."
+                f" This would require {requirement:.2f} MB,"
+                f" but only {self._memory_mb} MB is allowed."
+            )
+        if self._cache is None:
+            self._cache = self._alpha_matrix(self._current_size, size)
+        else:
+            self._cache = np.dstack([
+                self._cache,
+                self._alpha_matrix(self._current_size, size)
+            ])
+        self._current_size = size
+        return self._cache
+
+    def _single(self, vector: np.ndarray, alphas: np.ndarray):
+        r"""
+        Get the Q function (without the :math:`\pi` scaling factor) of a single
+        state vector.
+        """
+        return np.abs(np.dot(alphas, (self._g*0.5)*vector))**2
+
+    def __call__(self, state: Qobj):
+        """
+        Get the Husimi-Q function for the given state vector or density matrix,
+        over the coordinates used to initialise the class.  If called multiple
+        times, the states do not need to have the same dimensions, but none of
+        them can have tensor-product structure.
+        """
+        state = _qfunc_check_state(state)
+        alphas = self._alphas(state.shape[0])
```

Comment:
Something similar for the variable alphas? . 

# alphas represents a [n, m, k] sized numpy array of complex numbers with (n, m)
# representing the phase space grid and k representing the Hilbert space
# dimensions for the coherent state.